### DISCUSSION TITLE
아래 코드의 결과에 대해 설명해주세요.

### DISCUSSION THREAD

**Discussion by publdaze** [publdaze/js-study#87] [클로저]
- 아래 코드의 결과에 대해 설명해주세요.
  ```js
  function outer() {
    let getY;
    {
      const y = 6;
      getY = () => y;
    }
    console.log(getY());
    console.log(y);
  }
  
  outer();
  ```

**Comment by publdaze** [comment]
> - 아래 코드의 결과에 대해 설명해주세요.
>   ```js
>   function outer() {
>     let getY;
>     {
>       const y = 6;
>       getY = () => y;
>     }
>     console.log(getY());
>     console.log(y);
>   }
>   
>   outer();
>   ```

- outer가 호출되고, outer 함수의 실행 컨텍스트와 그에 바인딩되는 렉시컬 환경 생성
- outer 함수의 지역 변수로 getY가 환경레코드에 등록되어 관리됨
- 다음 코드 블럭이 실행되면 해당 코드 블럭의 렉시컬 환경이 생성됨
- 코드 블럭 내부가 실행 되면서 y 식별자가 코드 블럭의 환경레코드에 등록됨
- 다음 실행문은 getY에 함수가 할당되고 있음
- getY는 현재 렉시컬 환경에 존재하지 않기 때문에, 외부 렉시컬 환경에 대한 참조에 할당된 값인 outer 함수의 렉시컬 환경으로 검색을 이어감
- outer 함수의 렉시컬 환경에는 getY 식별자가 존재하므로 식별자가 결정되고, 함수 정의가 할당됨
- 이후 코드 블럭이 종료되는데, 내부 함수인 getY가 outer 함수의 식별자에 할당이 되면서 코드 블럭보다 긴 생명주기를 가지고
- getY에서 코드 블럭 렉시컬 환경에 저장된 y를 참조하기 때문에, getY 함수는 클로저라 할 수 있으며, 코드 블럭의 렉시컬 환경은 참조되고 있는 y 식별자에 의해 제거되지 않고 남아있음
- 이후 console.log 하위에서 getY가 호출되면, getY 함수에 대한 평가가 진행되고, y 식별자에 대한 검색이 진행됨
- y 식별자는 getY 함수에 대한 렉시컬 환경에서는 찾을 수 없고 외부 렉시컬 환경에서 찾을 수 있음. 따라서 y가 참조하고 있는 6이라는 값이 출력됨
- 두번째 console.log 하위에서 사용되는 y 식별자를 검색하면 스코프 체인을 통해 검색하는 과정중 해당 식별자를 찾을 수 없으므로 undefined가 출력됨

**Comment by pipisebastian** [comment]
```js
function outer() {
  let getY;
  {
    const y = 6;
    getY = () => y;
  }
  console.log(getY());
  console.log(y);
}

outer();

```

- `outer` 실행
    - `getY` → `outer` 함수의 렉시컬 환경에 저장
    - 코드 블록 `{}` 실행
        - `y` → `{}` 코드블록의 렉시컬 환경에 저장
        - `getY` 에 함수를 할당
    - 코드 블록 `{}` 종료
        - `getY` 는 `y` 를 참조하고 있음. `y` 는 코드블록의 렉시컬 환경에 있기에, 코드 블록이 종료되었지만 코드 블록의 렉시컬 환경은 살아있음.
    - `getY()` 실행
        - y 식별자를 참조할 수 있음 → 6 반환
    - `y` 출력
        - y 식별자를 참조하려 하지만, `{}` 코드블록의 식별자환경은 이미 실행이 끝난데다가, 외부에서는 접근할 수 없음. (`getY` 함수만 접근할 수 있음)
        - ⇒ reference 에러!
